src/System.Runtime.Extensions/src/System/Net/WebUtility.cs

```diff
@@ -331,30 +332,32 @@ private static byte[] UrlEncode(byte[] bytes, int offset, int count)
 
             // expand not 'safe' characters into %XX, spaces to +s
             byte[] expandedBytes = new byte[count + cUnsafe * 2];
-            int pos = 0;
+            GetEncodedBytes(bytes, offset, count,expandedBytes);
+            return expandedBytes;
+        }
 
-            for (int i = 0; i < count; i++)
+        private static void GetEncodedBytes(byte[] originalBytes, int startIndex, int count, byte[] newBytes)
+        {
+            int pos = 0;
+            for (int i = startIndex; i < startIndex + count; i++)
```

Comment:
I'm assuming the JIT can see that startIndex and count will never change and can avoid redoing the addition on each iteration.  If that's not the case, you could put this upper bound into a local before the loop and then just use that.